osconfig_tests/test_config/project_config.go

```diff
@@ -52,24 +52,45 @@ func GetProject(projectID string, testZones map[string]int) *Project {
 
 // GetZone gets a random zone that still has capacity.
 func (p *Project) GetZone() string {
-	p.mux.Lock()
-	defer p.mux.Unlock()
+	for {
+		p.mux.Lock()
+
+		zc := len(p.zoneIndices)
+		if zc == 0 {
+			p.mux.Unlock()
+			time.Sleep(10 * time.Second)
+			continue
+		}
+
+		// Pick a random zone.
+		zi := rand.Intn(zc)
+		z := p.zoneIndices[zi]
+
+		// Decrement the number of instances that this zone can host.
+		p.testZones[z]--
+		// Remove this zone from zoneIndices if it can't host any more instances.
+		if p.testZones[z] == 0 {
+			p.zoneIndices = append(p.zoneIndices[:zi], p.zoneIndices[zi+1:]...)
+		}
 
-	zc := len(p.zoneIndices)
-	if zc == 0 {
-		// TODO: return an error instead of stopping the process.
-		msg := "Not enough zone quota sepcified. Specify additional quota in `test_zones`."
-		fmt.Println(msg)
-		return msg
+		p.mux.Unlock()
+		return z
 	}
+}
 
-	zi := rand.Intn(zc)
-	z := p.zoneIndices[zi]
+// ReturnZone returns a zone so other tests can use it.
+func (p *Project) ReturnZone(z string) {
```

Comment:
GetZone and ReturnZone threw me off at first. Would Aquire/Release (following Go's semaphore semantics) be better?